RANSACApp/camera.py

```diff
@@ -1,65 +1,116 @@
-      if not self.capture_source.isOpened() or (self.config.capture_source != self.current_capture_source):
-        print(f"Capture source {self.config.capture_source} not found, retrying in 500ms")
-        sleep(0.5)
-        self.current_capture_source = self.config.capture_source
-        self.capture_source = cv2.VideoCapture(self.current_capture_source)
-        continue
-
-      # Assuming we can access our capture source, wait for another thread to request a capture.
-      # Cycle every so often to see if our cancellation token has fired. This basically uses a
-      # python event as a contextless, resettable one-shot channel.
-      if not self.capture_event.wait(timeout=0.02):
-        continue
-
-      try:
-        ret, image = self.capture_source.read()
-        if not ret:
-          self.capture_source.set(cv2.CAP_PROP_POS_FRAMES, 0)
-          raise RuntimeError("Problem while getting frame")
-        frame_number = self.capture_source.get(cv2.CAP_PROP_POS_FRAMES)
-        fps = self.capture_source.get(cv2.CAP_PROP_FPS)
+    def __init__(self, config: RansacConfig, camera_index: int, cancellation_event: "threading.Event",
+                 capture_event: "threading.Event", camera_status_outgoing: "queue.Queue[CameraState]",
+                 camera_output_outgoing: "queue.Queue"):
+        self.config = config
+        self.camera_index = camera_index
+        self.camera_address = config.capture_source
+        self.camera_status_outgoing = camera_status_outgoing
+        self.camera_output_outgoing = camera_output_outgoing
+        self.capture_event = capture_event
+        self.cancellation_event = cancellation_event
+        self.current_capture_source = config.capture_source
+        self.wired_camera: "cv2.VideoCapture" = cv2.VideoCapture(config.capture_source)
+        self.stream = None
+        self.stream_frame_number = 0
+        self.stream_bytes = bytes()
+        self.error_message = "Capture source {} not found, retrying in 500ms"
+
+    def set_output_queue(self, camera_output_outgoing: "queue.Queue"):
+        self.camera_output_outgoing = camera_output_outgoing
+
+    def run(self):
+        while True:
+            if self.cancellation_event.is_set():
+                print("Exiting capture thread")
+                return
+
+            # If things aren't open, retry until they are. Don't let read requests come in any earlier
+            # than this, otherwise we can deadlock ourselves.
+            if type(self.config.capture_source) == str and "http" in self.config.capture_source:
+                try:
+                    self.stream = requests.get(self.config.capture_source, stream=True)
+                except requests.exceptions.RequestException:
+                    print(self.error_message.format(self.config.capture_source))
+            else:
+                # so, they might have switched to a wired camera for some reason, no need to keep the stream running
+                self.cleanup_stream()
```

Comment:
It looks like you're missing a call to cleanup_stream() on the token based exit at the beginning of the loop. Does the program still exit ok if the stream isn't cleaned up?

Reply:
Ok awesome, didn't know if that would matter on shutdown and just wanted to make sure, thanks for updating that!